Editorial review: Document interest invokers (#42021)
mdn/content: files/en-us/web/api/popover_api/using_interest_invokers/index.md

```css live-sample___interest-invoker-styling
p:has(button:interest-source) button {
  interest-delay-start: 0s;
}
```

We've also set a {{cssxref("position-area")}} of `bottom` on the tooltip so that it appears below the button. This is possible because associating any popover with its interest invoker creates an implicit anchor reference between them (see [Popover anchor positioning](/en-US/docs/Web/API/Popover_API/Using#popover_anchor_positioning) for more details).

```css live-sample___interest-invoker-styling
#mytooltip {
  position-area: bottom;
}
```

Finally, we've used the `#mytooltip:interest-target` selector to set a dashed border on the popover when interest is shown.

```css live-sample___interest-invoker-styling
#mytooltip:interest-target {
  border-style: dashed;
}
```

This renders as follows:

{{embedlivesample("interest-invoker-styling", "100%", "150")}}

Try showing interest in the button (for example by hovering or focusing it) to observe the behavior described earlier.

> ## The interest invoker JavaScript API

Reviewer: to avoid starting with "The":

```suggestion
## JavaScript API for interest invokers
```

> Interest invokers have an associated JavaScript API, which can be used to query which element is being targeted by an interest invoker and run custom code in response to interest being shown or lost. The features of the API are as follows:

Reviewer: ```suggestion
Interest invokers have an associated JavaScript API that lets you query the element being targeted by an interest invoker and run custom code when interest is shown or lost. The main features of this API include:
```

- The {{domxref("HTMLButtonElement.interestForElement", "interestForElement")}} property, which is available on the {{domxref("HTMLButtonElement")}}, {{domxref("HTMLAnchorElement")}}, {{domxref("HTMLAreaElement")}}, and {{domxref("SVGAElement")}} interfaces, and returns a reference to the element being targeted by the interest invoker. This will be the element whose `id` is referenced in the equivalent HTML or SVG interest invoker element's `interestfor` attribute.

> - The {{domxref("HTMLElement.interest_event", "interest")}} and {{domxref("HTMLElement.loseinterest_event", "loseinterest")}} events, which are fired on an interest invoker's target element when interest is shown and lost, respectively, allowing custom code to be run in response.

Reviewer: ```suggestion
- The {{domxref("HTMLElement.interest_event", "interest")}} and {{domxref("HTMLElement.loseinterest_event", "loseinterest")}} events, which are fired on an interest invoker's target element when interest is shown and lost, respectively. You can listen for these events to run custom code in response.
```

}

#mytooltip {
  position-area: bottom;
}
```

In our JavaScript, we grab references to our popover and the three links. We then loop through the links, setting the {{domxref("HTMLAnchorElement.interestForElement", "interestForElement")}} property of each one to equal the reference to our popover. This sets up the interest invoker/target relationship between the popover and each of the three links programmatically.

```js live-sample___interest-invoker-api
const tooltip = document.getElementById("mytooltip");
const links = document.querySelectorAll("a");
links.forEach((link) => (link.interestForElement = tooltip));
```

We then attach `interest` and `loseinterest` event handlers to the popover. When interest is shown on one of the links, we update the popover's text content to include the text content of the link (retrieved via the event object's `source` property), so you can see which link caused the popover to appear. When interest is lost, we append an asterisk to the `source` element's text content, so you can see how many times each one has been used to show interest.

```js live-sample___interest-invoker-api
tooltip.addEventListener("interest", (e) => {

>   tooltip.textContent = `Interest gained via ${e.source.textContent}`;

Reviewer: ```suggestion
  tooltip.textContent = `Interest shown on ${e.source.textContent}`;
```

In addition to activation-based popovers, there is also a common need to display a popover when a control element is hovered or focused – interactions that indicate user interest. For example, many social and community sites let users hover over a link to a person's profile page or a group page to show a popover with more information, helping users decide whether they want to visit the page. Such popovers may also contain quick actions, such as "Subscribe to group" or "Follow", allowing users to take an action without losing their current context.

Interest invokers enable the browser to provide this functionality in a consistent and accessible way, without requiring JavaScript. The browser determines when a user is showing interest in an element, and therefore, when an action should be taken. "Showing interest" generally occurs when a user hovers, focuses, or long-presses the element (the exact nature of "interest" may vary across browsers), and "losing interest" generally happens when the user stops interacting with the element.

> The browser also fires events when interest is gained or lost, so you can run custom code in response. In addition, the feature includes CSS properties and selectors for styling based on interest.

Reviewer: ```suggestion
The browser also fires events when interest is gained or lost, so you can run custom code in response. In addition, this feature includes CSS properties and selectors for styling elements based on the interest status.
```

if (!supported) {
  document.querySelector("html").classList.add("no-interest-invokers");
}
```

```html live-sample___basic-interest-invoker
<p>Some text with <a href="#" interestfor="mypopover">a link</a>.</p>
<p id="mypopover" popover>A short preview with some quick info</p>
```

Setting the `popover` attribute on the target element causes it to be hidden (via {{cssxref("display", "display: none")}}) and centered on the screen. Showing interest in the invoker element (the link) causes the popover to appear.

This renders as follows. Try interacting with the link:

{{embedlivesample("basic-interest-invoker", "100%", "150")}}

Note how the popover appears when the link is hovered, focused, or long-pressed, and disappears when the interaction stops. If the link is activated (for example, with a mouse click), it functions normally; though in this example the link doesn't go anywhere.

The value of the `popover` attribute doesn't affect the popover's behavior in this example. However, it becomes important when you combine interest invoker popovers with regular popovers, as shown in the next section.

## Interest invoker interaction with other popovers

> It is possible to combine interest invokers with regular popovers on the same control. In this example, we have a {{htmlelement("button")}} element that is set to be an interest invoker via the `interestfor` attribute, meaning it will show a tooltip when interest is shown in it. However, when it is clicked, it will show/hide a different popover, referenced inside the `commandfor` attribute. The [`command`](/en-US/docs/Web/HTML/Reference/Elements/button#command) attribute is set to `toggle-popover` so that it can be pressed multiple times to toggle the popover between showing and hidden states.

Reviewer: ```suggestion
You can combine interest invokers with regular popovers on the same control element. In the following example, a {{htmlelement("button")}} element is set up as an interest invoker using the `interestfor` attribute, meaning it will show a tooltip when the user shows interest in it. If the button is clicked, it will show or hide a different popover referenced by the `commandfor` attribute. The [`command`](/en-US/docs/Web/HTML/Reference/Elements/button#command) attribute is set to `toggle-popover`, allowing the button to be pressed multiple times to toggle the popover between its shown and hidden states.
```

```css hidden live-sample___interest-invoker-popover-interaction
#mytooltip {
  position-area: right;
}

#myinfobox {
  position-area: bottom;
}
```

```html live-sample___interest-invoker-popover-interaction
<p>
  Some basic text with
  <button
    interestfor="mytooltip"
    commandfor="myinfobox"
    command="toggle-popover">
    a button
  </button>

>   inside it.

Reviewer: ```suggestion
```

</p>
<p id="mytooltip" popover="hint">A hover toolip</p>
<p id="myinfobox" popover>
  An infobox that also contains some control buttons<br />
  <button>Button 1</button> <button>Button 2</button>
</p>
```

This renders as follows:

{{embedlivesample("interest-invoker-popover-interaction", "100%", "150")}}

> In this case, you can show interest in the button to reveal the hover tooltip, and then press the button to reveal the infobox. Note that the `popover` values matter here — we've got the tooltip popover set with [`popover="hint"`](/en-US/docs/Web/API/Popover_API/Using#using_hint_popover_state), whereas the infobox is set to just `popover` (which is equivalent to `popover="auto"`). This allows the tooltip to be shown when the infobox is already shown, without hiding it. If both popovers were set to `auto`, you wouldn't be able to see both at the same time. In a UI, it is useful to be able to look at multiple tooltips without hiding parts of the UI that you've previously opened.

Reviewer: ```suggestion
You can show interest in the button by hovering, which displays the tooltip, and then click the button to reveal the infobox. Note that the `popover` values matter here — the tooltip popover is set to [`popover="hint"`](/en-US/docs/Web/API/Popover_API/Using#using_hint_popover_state), whereas the infobox is set to just `popover` (equivalent to `popover="auto"`). This allows the tooltip to stay visible even when the infobox is shown. If both popovers were set to `auto`, you wouldn't be able to see both the tooltip and the infobox at the same time. In a UI, it is useful to be able to view multiple tooltips without hiding parts of the UI that you've already opened.
```

- The [`interest-delay`](/en-US/docs/Web/CSS/Reference/Properties/interest-delay) shorthand property and its related [`interest-delay-start`](/en-US/docs/Web/CSS/Reference/Properties/interest-delay-start) and [`interest-delay-end`](/en-US/docs/Web/CSS/Reference/Properties/interest-delay-end) longhands, which can be set to add a delay between the user showing/losing interest and the interest change being acted on by the browser — for example, the popover being shown/hidden.
- The [`:interest-source`](/en-US/docs/Web/CSS/Reference/Selectors/:interest-source) and [`:interest-target`](/en-US/docs/Web/CSS/Reference/Selectors/:interest-target) pseudo-classes, which can be used to apply styles to the interest invoker and its associated target element, respectively, only when interest is being shown.

Let's look at a simple example that demonstrates their use.

We specify two buttons and a tooltip, which is shown/hidden when the user shows/loses interest in the `<button>`:

```html live-sample___interest-invoker-styling
<p>
  <button interestfor="mytooltip">Button 1</button>
  <button interestfor="mytooltip">Button 2</button>
</p>
<p id="mytooltip" popover="hint">A hover toolip</p>
```

> In the CSS, we set an `interest-delay` of `1s 2s` on the `<button>` — this creates a delay of 1 second between the user showing interest and the popover appearing, and a delay of 2 seconds between the user losing interest and the popover disappearing. We also use the `button:interest-source` selector to set an `orange` {{cssxref("background-color")}} on the `<button>` only when interest is being shown.

Reviewer: ```suggestion
In the CSS, we've set an `interest-delay` of `1s 2s` on the `<button>` — this creates a delay of 1 second before the tooltip appears when the user shows interest, and a delay of 2 seconds before it disappears when the user loses interest. We've also used the `button:interest-source` selector to set an `orange` {{cssxref("background-color")}} on the `<button>` when interest is shown.
```

}
```

Next, we combine the `:interest-source` pseudo-class with the {{cssxref(":has()")}} pseudo-class to create a ruleset that applies `interest-delay-start: 0s` to all buttons inside the paragraph, only if the paragraph contains a button that interest has been shown on (that is, has been selected by `button:interest-source`). This is a useful technique — having the popover appear as soon as interest is shown on any button would create an annoying user experience, but after the user has shown interest in a button, it is convenient for them to be able to quickly move between different popovers.

```css live-sample___interest-invoker-styling
p:has(button:interest-source) button {
  interest-delay-start: 0s;
}
```

Next, we give the popover a {{cssxref("position-area")}} of `bottom`; this causes the popover to be positioned below the `<button>`. This is possible because associating any kind of popover with its invoker creates an implicit anchor reference between the two (see [Popover anchor positioning](/en-US/docs/Web/API/Popover_API/Using#popover_anchor_positioning) for more details).

```css live-sample___interest-invoker-styling
#mytooltip {
  position-area: bottom;
}
```

> Finally, we use the `#mytooltip:interest-target` selector to set a dashed border on the popover only when interest is being shown.

Reviewer: ```suggestion
Finally, we use the `#mytooltip:interest-target` selector to set a dashed border on the popover when interest is shown.
```

- `revertStyle()`: When interest is lost, the style panel is reverted back to its previous style by setting the class stored in `prevStyle` as its `className`.
- `setStyle()`: When a button is actually clicked, its `className` is set as the style panel's `className`. We also set `prevStyle` to equal the clicked button's `className` so that, next time a style is previewed, it will revert back to the _new_ previously-set style.

```js live-sample___non-popover
function sampleStyle(e) {
  e.target.className = e.source.className;
}

function revertStyle(e) {
  e.target.className = prevStyle;
}

function setStyle(e) {
  stylePanel.className = e.target.className;
  prevStyle = e.target.className;
}
```

### Result

This renders as follows:

{{embedlivesample("non-popover", "100%", "260")}}

Try hovering/focusing the buttons to preview the styles on the style panel, and activating them to apply styles permanently. Note that setting the styles will still work on non-supporting browsers, even though the "preview" functionality doesn't.

> ## Link preview popover example

Reviewer: would this make sense?
```suggestion
## Using interest invokers for creating preview popovers
```

As mentioned earlier, a very common use case for interest invokers is to progressively enhance links to show some preview information about the link target, such as the name, bio, and location of a user on a link to their profile page, or quick actions like subscribing to a group on a link to the group's homepage. Such popovers are convenient, as they often give the user what they need without forcing them to lose context by making an additional navigation.

Let's look at how to implement a preview popover using interest invokers.

### HTML

In the markup, we include a link to my GitHub profile inside a short paragraph, and a `<div>` containing a limited user profile with a fake "Follow" button. The link has `interestfor` set to the `id` of the user profile, and the user profile has a `popover` attribute set to turn it into a popover element (this hides it by default).

```html live-sample___link-preview-popover
<p>
  I think
  <a
    href="https://github.com/chrisdavidmills/"
    interestfor="user-info"
    target="_blank">
    @chrisdavidmills
  </a>

>   should know about this.

Reviewer: how about:
- `"Here's an example from @chrisdavidmills."`
or 
- `"Check out what @chrisdavidmills has been working on."`

Author: I'm not sure this is particularly needed, and it puts it further out of sync with the version in `dom-examples`. I'm going to leave this one.

</p>

<div id="user-info" popover="hint">
  <div class="wrapper">
    <img src="chris-mills.jpg" alt="chris mills" />
    <section>
      <p><strong>Chris Mills</strong></p>
      <p>
        Independent tech writer and web technology tinkerer, working on MDN on
        behalf of Google and Mozilla. A11y and open standards advocate. Heavy

>         metal drummer. <button>Follow</button>

Reviewer: ```suggestion
        Independent technical writer and web technology tinkerer.
        Working on MDN for Google and Mozilla.
        A11y and open standards advocate.
        Heavy-metal drummer. <button>Follow</button>
```

      </p>
      <p>🌍 Greenfield, UK</p>
    </section>
  </div>
</div>
```

### CSS

> We begin our styles by setting an {{cssxref("interest-delay-start")}} value of `1s` on the link (the `a[interestfor]` selector is useful for selecting only links that are interest invokers). We are setting this to increase the delay between a user pausing over a link and the popover appearing. It could be annoying to have popovers appearing too quickly on a dense, link-heavy page.

Reviewer: ```suggestion
We start by setting an {{cssxref("interest-delay-start")}} value of `1s` on the link (the `a[interestfor]` selector is useful for selecting only the links that are interest invokers). This creates a slight delay before the preview popover appears, otherwise it can get annoying with popovers appearing too quickly on a dense, link-heavy page.
```

```css hidden live-sample___link-preview-popover
html {
  font-family: sans-serif;
}

* {
  box-sizing: border-box;
}
```

```css live-sample___link-preview-popover
a[interestfor] {
  interest-delay-start: 1s;
}
```

> Our next CSS rule sets a `position-area` value of `bottom right` on our popover so that it appears at the bottom-right of the link when interest is gained. (We've hidden most of the popover styling for brevity.)

Reviewer: ```suggestion
Next, we set a `position-area` value of `bottom right` on the popover so that it appears at the lower-right edge of the link when interest is shown. (The rest of the popover styling is hidden here for brevity.)
```
